marshmallow_recipe/fields.py

```diff
@@ -581,16 +585,81 @@ def raw_field(
     )
 
 
+def union_field(
+    fields: list[m.fields.Field],
+    *,
+    required: bool,
+    allow_none: bool,
+    default: Any = dataclasses.MISSING,
+    name: str | None = None,
+    validate: ValidationFunc | collections.abc.Sequence[ValidationFunc] | None = None,
+    **_: Any,
+) -> m.fields.Field:
+    if default is m.missing:
+        return UnionField(
+            fields=fields,
+            allow_none=allow_none,
+            validate=validate,
+            **default_fields(m.missing),
+            **data_key_fields(name),
+        )
+
+    if required:
+        if default is None:
+            raise ValueError("Default value cannot be none")
+        return UnionField(
+            fields=fields,
+            required=True,
+            allow_none=allow_none,
+            validate=validate,
+            **data_key_fields(name),
+        )
+
+    return UnionField(
+        fields=fields,
+        allow_none=allow_none,
+        validate=validate,
+        **(default_fields(None) if default is dataclasses.MISSING else {}),
+        **data_key_fields(name),
+    )
+
+
 DateTimeField: type[m.fields.DateTime]
 EnumField: type[m.fields.Field]
 DictField: type[m.fields.Field]
 SetField: type[m.fields.List]
 FrozenSetField: type[m.fields.List]
 TupleField: type[m.fields.List]
 StrField: type[m.fields.Str]
+UnionField: type[m.fields.Field]
 
 if _MARSHMALLOW_VERSION_MAJOR >= 3:
 
+    def validate_value_v3(
+        field: TField,
+        type_guards: type | tuple[type, ...] | None = None,
+        is_dataclass: bool = False,
+    ) -> TField:
+        fail_key = "invalid" if "invalid" in field.default_error_messages else "validator_failed"
+
+        old = field._serialize  # type: ignore
+
+        def _serialize_with_validate(self: TField, value: Any, attr: Any, obj: Any, **kwargs: Any) -> Any:
+            if value is not None and (
+                type_guards is not None
+                and not isinstance(value, type_guards)
+                or is_dataclass
+                and not dataclasses.is_dataclass(value)
```

Comment:
do we actually need to do an extra check for dataclass? as I see method call with `is_dataclass=True` always has `type_guards` argument which contains dataclass type and `isinstance(value, type_guards)` already should cover all we need